File: relay-backend/app/schemas/call.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from uuid import UUID
from enum import Enum

class CallStatus(str, Enum):
    SCHEDULED = "scheduled"
    QUEUED = "queued"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"

class CallPriority(str, Enum):
    LOW = "low"
    NORMAL = "normal"
    HIGH = "high"

class CallOutcome(str, Enum):
    SUCCESSFUL = "successful"
    NO_ANSWER = "no_answer"
    BUSY = "busy"
    VOICEMAIL = "voicemail"
    FAILED = "failed"
    HUMAN_HANGUP = "human_hangup"

class ScheduleCallRequest(BaseModel):
    contact_id: UUID
    template_id: UUID
    scheduled_at: datetime
    priority: CallPriority = CallPriority.NORMAL
    max_retries: int = Field(default=3, ge=0, le=10)
    metadata: dict = Field(default_factory=dict)

class ScheduledCallResponse(BaseModel):
    id: UUID
    user_id: UUID
    contact_id: UUID
    template_id: UUID
    status: CallStatus
    priority: CallPriority
    scheduled_at: datetime
    retry_count: int
    max_retries: int
    metadata: dict
    created_at: datetime

class CallRecordResponse(BaseModel):
    id: UUID
    scheduled_call_id: UUID
    provider_call_id: Optional[str]
    started_at: Optional[datetime]
    ended_at: Optional[datetime]
    duration_seconds: Optional[int]
    outcome: Optional[CallOutcome]
    recording_url: Optional[str]
    transcript: list[dict]
    created_at: datetime

class CallDetailResponse(BaseModel):
    call: ScheduledCallResponse
    records: list[CallRecordResponse] = Field(default_factory=list)
    extracted_data: Optional[dict] = None

File: relay-frontend/types/call.ts
export type CallStatus = 'scheduled' | 'queued' | 'in_progress' | 'completed' | 'failed' | 'cancelled'
export type CallPriority = 'low' | 'normal' | 'high'
export type CallOutcome = 'successful' | 'no_answer' | 'busy' | 'voicemail' | 'failed' | 'human_hangup'

export interface ScheduledCall {
  id: string
  user_id: string
  contact_id: string
  template_id: string
  status: CallStatus
  priority: CallPriority
  scheduled_at: string
  retry_count: number
  max_retries: number
  metadata: Record<string, unknown>
  created_at: string
}

export interface CallRecord {
  id: string
  scheduled_call_id: string
  provider_call_id: string | null
  started_at: string | null
  ended_at: string | null
  duration_seconds: number | null
  outcome: CallOutcome | null
  recording_url: string | null
  transcript: Array<{ speaker: string; text: string; timestamp: number }>
  created_at: string
}

export interface ExtractedAnswer {
  value: string | null
  confidence: number
  source_quote?: string
}

export interface ExtractedData {
  structured_data: {
    answers: Record<string, ExtractedAnswer>
    summary: string
    sentiment: 'positive' | 'neutral' | 'negative' | 'unknown'
    key_points: string[]
    follow_up_needed: boolean
    follow_up_reason?: string
  }
  confidence_score: number
  extraction_model: string
  created_at: string
}

export interface CallDetail {
  call: ScheduledCall
  records: CallRecord[]
  extracted_data: ExtractedData | null
}

export interface ScheduleCallInput {
  contact_id: string
  template_id: string
  scheduled_at: string
  priority?: CallPriority
  max_retries?: number
  metadata?: Record<string, unknown>
}
